controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,20 +124,296 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
+
+		// storageClassClaimStorageType is the storage type of the StorageClassClaim
+		var storageClassClaimStorageType providerclient.StorageType
+		switch r.storageClassClaim.Spec.Type {
+		case "blockpool":
+			storageClassClaimStorageType = FulfillStorageClassClaimRequestBlockpool
+		case "sharedfilesystem":
+			storageClassClaimStorageType = FulfillStorageClassClaimRequestSharedfilesystem
+		}
+
+		// Call the `FulfillStorageClassClaim` service on the provider server with StorageClassClaim as a request message.
+		_, err = providerClient.FulfillStorageClassClaim(
+			r.ctx,
+			r.storageCluster.Status.ExternalStorage.ConsumerID,
+			r.storageClassClaim.Name,
+			r.storageClassClaim.Spec.EncryptionMethod,
+			storageClassClaimStorageType,
+		)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to initiate fulfillment of StorageClassClaim: %v", err)
+		}
+
+		// Configuration phase.
+		r.storageClassClaim.Status.Phase = ocsv1alpha1.StorageClassClaimConfiguring
+
+		// Call the `GetStorageClassClaimConfig` service on the provider server with StorageClassClaim as a request message.
+		response, err := providerClient.GetStorageClassClaimConfig(
+			r.ctx,
+			r.storageCluster.Status.ExternalStorage.ConsumerID,
+			r.storageClassClaim.Name,
+		)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to get StorageClassClaim config: %v", err)
+		}
+		resources := response.ExternalResource
+		if resources == nil {
+			return reconcile.Result{}, fmt.Errorf("no configuration data recieved")
+		}
+
+		// Go over the received objects and operate on them accordingly.
+		for _, resource := range resources {
+			data := map[string]string{}
+			err = json.Unmarshal(resource.Data, &data)
+			if err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to unmarshal StorageClassClaim configuration response: %v", err)
+			}
+
+			switch resource.Kind {
+			case "Secret":
+				secret := &corev1.Secret{}
+				secret.Name = resource.Name
+				secret.Namespace = r.storageClassClaim.Namespace
+				_, err = controllerutil.CreateOrUpdate(r.ctx, r.Client, secret, func() error {
+					err := r.own(secret)
+					if err != nil {
+						return fmt.Errorf("failed to own Secret: %v", err)
+					}
+					m := map[string][]byte{}
+					for k, v := range data {
+						m[k] = []byte(v)
+					}
+					secret.Data = m
+					return nil
+				})
+				if err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to create or update secret %v: %s", secret, err)
+				}
+			case "StorageClass":
+				var storageClass *storagev1.StorageClass
+				if resource.Name == "cephfs" {
+					storageClass = r.getCephFSStorageClass(data)
+				} else if resource.Name == "ceph-rbd" {
+					storageClass = r.getCephRBDStorageClass(data)
+				}
+				err = r.createOrReplaceStorageClass(storageClass)
+				if err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to create or update StorageClass: %s", err)
+				}
```

Comment:
This can be done by deleting existing claim and creating new claim.

"Replace StorageClass" is a scary function to automate. We should not allow changes to Claim in such a way that we end up replacing StorageClass.

IMO this should be updated to "create StorageClass or fail if replace is required".

@Madhu-1 do you think it makes sense?

Reply:
Why this is required?

Reply:
Do you mean my change or Umanga's suggestion?

Reply:
I mean you dont need to Delete storageclass or update it right? if it's present just use it

Reply:
If we allow updating storageclassclaim we might need to delete and recreate the storageclass if there is a parameter change https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/storagecluster/storageclasses.go#L137-L167

Reply:
I agree with @umangachapagain  and @Madhu-1 , a storage class claim should be considered spec immutable after creation. 

But I think the correct way is to validate that the spec is not conflicting with the actual storageclass at the start of the reconcile (if a storage class already exists)  as part  a`validating` phase. This way we exist early before running any kind of API server calls, or trying to fullil anything.